### ISSUE TITLE
oci8 num_rows breaks transactions

### ISSUE THREAD

**Issue by a-krebs** [bcit-ci/CodeIgniter#696]
This could be related to #182 and #682:

in oci8_result, the function num_rows runs queries again before returning (note #682 is still open, so I have changed the return call to be return $this->num_rows)

``` php
    public function num_rows()
    {
        if ($this->num_rows === 0 && count($this->result_array()) > 0)
        {
            $this->num_rows = count($this->result_array());
            @oci_execute($this->stmt_id);

            if ($this->curs_id)
            {
                @oci_execute($this->curs_id);
            }
        }

        return $this->num_rows(); // is currently "return $rowcount" which is a bug as noted above
    }
```

When these queries are run, they are committed (and not treated as part of the transaction). This issue is present even if num_rows() is never called directly, so I have a feeling num_rows() is called from somewhere else in the database library (either somewhere in $this->db->query(), or in $resultSet->result()).

Thus, num_rows() will break a transaction lock. I noticed this when trying to do update BLOB columns in my database (which require row lock).

The code I was trying to run is:

``` php
$empty_blob_set = $this->db->query('SELECT photo FROM images WHERE photo_id = ?', $photoId);
foreach($empty_blob_set->result_array() as $row) {
    $blob = $row['PHOTO'];
    $blob->write($imageAsString);
}
```

What is the purpose of running the

``` php
@oci_execute($this->stmt_id);
if ($this->curs_id)
{
    @oci_execute($this->curs_id);
}
```

block? If I change the @oci_execute to use OCI_DEFAULT (and thus not commit), the issue appears to go away (but removing the @oci_execute entirely means empty results are returned by $this->db->query() calls). I don't know what side-effects OCI_DEFAULT has, as I'm not sure why the @oci_execute calls are there, so this isn't really a fix.
# EDIT

After digging through the code for a bit, it appears that in DB_driver.php, line 386, after the result ojbect is instantiated, the call to num_rows() forces a call (via oci8_result.php line 45) to `result_array()`, which calls `_fetch_assoc()` to build the array.
This means that the cursor is moved to the end of the OCI result object with `oci_feth_assoc()`.
Then when, later on, `$resultSet->results()` is called, a results object cannot be build because the `oci_fetch_object()` call in `result_object()` and then `_fetch_object()` returns FALSE, since no more rows are available in the OCI result object.

This can be verified by using only `$resultSet->result_array()`. Since this array is already built, it will work as intended, while `$resultSet->result()`, which uses a result object, will be broken since the result object cannot be assembled.

In conclusion, this block in oci8_results.php num_rows function:

``` php
@oci_execute($this->stmt_id);
if ($this->curs_id)
{
    @oci_execute($this->curs_id);
}
```

exists only to move the cursor back to the beginning of the result set. I therefore propose that calling the oci_execute methods with OCI_NO_AUTO_COMMIT (or equivalently OCI_DEFAULT) is a suitable workaround, since it's acceptable to always not commit this query, since it will only ever apply to SELECT statements and only needs to be committed when other changes are made (and will be committed in that case, when the change is committed).

I'll do some testing to make sure this all works, and then write a fix.
